src/script/builder/tir/block_frame.cc

```diff
@@ -39,24 +41,115 @@ BlockFrame Block_(String name) {
+  BlockFrame frame = Builder::Current()->FindFrame<BlockFrame>().value();
+  frame->init = init.value_or(tvm::tir::Evaluate(0));
+}
+
+tvm::tir::BufferRegion BufferRegionFromLoad(tvm::tir::BufferLoad buffer_load) {
+  using namespace tvm::tir;
+  Array<Range> ranges;
+  for (const PrimExpr& index : buffer_load->indices) {
+    ranges.push_back(Range::FromMinExtent(index, 1));
+  }
+  return BufferRegion(buffer_load->buffer, ranges);
+}
+
+void BlockWhere(PrimExpr predicate) {
+  BlockFrame frame = Builder::Current()->FindFrame<BlockFrame>().value();
+  if (frame->predicate.defined()) {
+    LOG(FATAL) << "Duplicate block predicate declaration, previous one is "
+               << frame->predicate.value();
+  }
+  frame->predicate = predicate;
+}
+
+void BlockReads(Array<ObjectRef> buffer_slices) {
+  using namespace tvm::tir;
+  BlockFrame frame = Builder::Current()->FindFrame<BlockFrame>().value();
+  if (frame->reads.size()) {
+    LOG(FATAL) << "Duplicate read region declaration, previous one is " << frame->reads;
+  }
+  for (const ObjectRef& obj : buffer_slices) {
+    if (const auto* buffer_region = obj.as<BufferRegionNode>()) {
+      frame->reads.push_back(GetRef<BufferRegion>(buffer_region));
+    } else if (const auto* buffer_load = obj.as<BufferLoadNode>()) {
+      frame->reads.push_back(BufferRegionFromLoad(GetRef<BufferLoad>(buffer_load)));
+    } else {
+      LOG(FATAL) << "Invalid type for buffer reads.";
+    }
+  }
+}
+
+void BlockWrites(Array<ObjectRef> buffer_slices) {
+  using namespace tvm::tir;
+  BlockFrame frame = Builder::Current()->FindFrame<BlockFrame>().value();
+  if (frame->writes.size()) {
+    LOG(FATAL) << "Duplicate write region declaration, previous one is " << frame->writes;
+  }
+  for (const ObjectRef& obj : buffer_slices) {
+    if (const auto* buffer_region = obj.as<BufferRegionNode>()) {
+      frame->writes.push_back(GetRef<BufferRegion>(buffer_region));
+    } else if (const auto* buffer_load = obj.as<BufferLoadNode>()) {
+      frame->writes.push_back(BufferRegionFromLoad(GetRef<BufferLoad>(buffer_load)));
+    } else {
+      LOG(FATAL) << "Invalid type for buffer writes.";
+    }
+  }
+}
+
+void BlockAttrs(Map<String, ObjectRef> attrs) {
+  BlockFrame frame = Builder::Current()->FindFrame<BlockFrame>().value();
+  frame->annotations = attrs;
 }
 
+tvm::tir::Buffer AllocBuffer(Array<PrimExpr> shape, DataType dtype, Optional<tvm::tir::Var> data,
+                             Array<PrimExpr> strides, PrimExpr elem_offset, String storage_scope,
+                             int align, int offset_factor, String buffer_type_str,
+                             Array<IntImm> axis_separators, Span span) {
+  using namespace tvm::tir;
+  Buffer buffer = DeclBuffer(shape, dtype, "", data, strides, elem_offset, storage_scope, align,
+                             offset_factor, buffer_type_str, axis_separators, span);
+  BlockFrame frame = Builder::Current()->FindFrame<BlockFrame>().value();
```

Comment:
I suggest that we don't over-use FindFrame for the following reason: we should rule out the possibility that `T.alloc_buffer` is called in a weird scope, for example:

```python
with T.block("..."):
  for i in range(10):
    T.alloc_buffer(...)  # <=== we should disallow this case
```

In this case, we should add a method:

```C++
inline Optional<TFrame> BuilderNode::GetLastFrame() const {
  if (frames is not empty) and (frames.back() is TFrameNode) {
    return downcast<TFrame> frames.back(); 
  }
  return NullOpt;
}
```